@seberg @ngoldbaum I've reworked this PR to use the buffer approach suggested by @ngoldbaum. Unfortunately, the tests all fail with the same kind of error:

```
[gw1] linux -- Python 3.11.6 /opt/hostedtoolcache/Python/3.11.6/x64/bin/python

    def test_find_access_past_buffer():
        # This checks that no read past the string buffer occurs in
        # string_fastsearch.h. The READC macro is there to check this.
        # To see it in action, you can redefine READC to just read the
        # i'th character of the buffer and this test will produce an
        # 'Invalid read' if run under valgrind.
        arr = np.array([b'abcd', b'ebcd'])
>       result = np._core.umath.find(arr, b'cde', 0, np.iinfo(np.int64).max)
E       numpy._core._exceptions._UFuncNoLoopError: ufunc 'find' did not contain a loop with signature matching types (<class 'numpy.dtypes.BytesDType'>, <class 'numpy.dtypes.BytesDType'>, <class 'numpy._IntegerAbstractDType'>, <class 'numpy._IntegerAbstractDType'>) -> None
```

This seems completely unrelated to the change I did in the last commits and I can't reproduce it locally either on a Ubuntu machine or an M1 macbook. Any ideas what might be causing it?

--- a/benchmarks/benchmarks/bench_core.py
+++ b/benchmarks/benchmarks/bench_core.py
@@ -254,6 +254,9 @@ class NumPyChar(Benchmark):
         self.A = np.array([100*'x', 100*'y'])
         self.B = np.array(1000 * ['aa'])
 
+        self.C = np.array([100*'x' + 'z', 100*'y' + 'z' + 'y', 100*'x'])
+        self.D = np.array(1000 * ['ab'] + 1000 * ['ac'])
+
     def time_isalpha_small_list_big_string(self):
         np.char.isalpha(self.A)
 
@@ -265,3 +268,9 @@ class NumPyChar(Benchmark):
 
     def time_add_big_list_small_string(self):
         np.char.add(self.B, self.B)
+
+    def time_find_small_list_big_string(self):
+        np.char.find(self.C, 'z')
+
+    def time_find_big_list_small_string(self):
+        np.char.find(self.D, 'b')

--- a/doc/release/upcoming_changes/24868.new_feature.rst
+++ b/doc/release/upcoming_changes/24868.new_feature.rst
@@ -1,0 +1,5 @@
+Add ``np.core.umath.find`` and ``np.core.umath.rfind`` UFuncs
+-------------------------------------------------------------
+Add two ``find`` and ``rfind`` UFuncs that operate on unicode or byte strings
+and are used in ``np.char``. They operate similar to ``str.find`` and
+``str.rfind``.

--- a/numpy/_core/code_generators/generate_umath.py
+++ b/numpy/_core/code_generators/generate_umath.py
@@ -1146,6 +1146,16 @@ defdict = {
     Ufunc(1, 1, False_,
           docstrings.get('numpy._core.umath.isalpha'),
           None,
+          ),
+'find':
+    Ufunc(4, 1, None,
+          docstrings.get('numpy.core.umath.find'),
+          None,
+          ),
+'rfind':
+    Ufunc(4, 1, None,
+          docstrings.get('numpy.core.umath.rfind'),
+          None,
           )
 }
 

--- a/numpy/_core/code_generators/ufunc_docstrings.py
+++ b/numpy/_core/code_generators/ufunc_docstrings.py
@@ -4276,3 +4276,77 @@ add_newdoc('numpy._core.umath', 'isalpha',
     --------
     str.isalpha
     """)
+
+add_newdoc('numpy.core.umath', 'find',
+    """
+    For each element, return the lowest index in the string where
+    substring `x2` is found, such that `sub` is contained in the
+    range [`x3`, `x4`].
+
+    Parameters
+    ----------
+    x1 : array_like, with ``bytes_`` or ``unicode_`` dtype
+
+    x2 : array_like, with ``bytes_`` or ``unicode_`` dtype
+
+    x3 : array_like, with ``int_`` dtype
+
+    x4 : array_like, with ``int_`` dtype
+        $PARAMS
+
+    `x3` and `x4` are interpreted as in slice notation.
+
+    Returns
+    -------
+    y : ndarray
+        Output array of ints
+        $OUT_SCALAR_2
+
+    See Also
+    --------
+    str.find
+
+    Examples
+    --------
+    >>> a = np.array(["NumPy is a Python library"])
+    >>> np.char.find(a, "Python", 0, None)
+    array([11])
+
+    """)
+
+add_newdoc('numpy.core.umath', 'rfind',
+    """
+    For each element, return the highest index in the string where
+    substring `x2` is found, such that `sub` is contained in the
+    range [`x3`, `x4`].
+
+    Parameters
+    ----------
+    x1 : array_like, with ``bytes_`` or ``unicode_`` dtype
+
+    x2 : array_like, with ``bytes_`` or ``unicode_`` dtype
+
+    x3 : array_like, with ``int_`` dtype
+
+    x4 : array_like, with ``int_`` dtype
+        $PARAMS
+
+    `x3` and `x4` are interpreted as in slice notation.
+
+    Returns
+    -------
+    y : ndarray
+        Output array of ints
+        $OUT_SCALAR_2
+
+    See Also
+    --------
+    str.find
+
+    Examples
+    --------
+    >>> a = np.array(["NumPy is a Python library"])
+    >>> np.char.find(a, "Python", 0, None)
+    array([11])
+
+    """)

--- a/numpy/_core/defchararray.py
+++ b/numpy/_core/defchararray.py
@@ -773,9 +773,8 @@ def find(a, sub, start=0, end=None):
     array([11])
 
     """
-    return _vec_string(
-        a, int_, 'find', [sub, start] + _clean_args(end))
-
+    end = end if end is not None else numpy.iinfo(numpy.int64).max
+    return numpy._core.umath.find(a, sub, start, end)
 
 @array_function_dispatch(_count_dispatcher)
 def index(a, sub, start=0, end=None):
@@ -1303,8 +1302,8 @@ def rfind(a, sub, start=0, end=None):
     str.rfind
 
     """
-    return _vec_string(
-        a, int_, 'rfind', [sub, start] + _clean_args(end))
+    end = end if end is not None else numpy.iinfo(numpy.int64).max
+    return numpy._core.umath.rfind(a, sub, start, end)
 
 
 @array_function_dispatch(_count_dispatcher)

--- a/numpy/_core/src/umath/dispatching.c
+++ b/numpy/_core/src/umath/dispatching.c
@@ -952,7 +952,8 @@ promote_and_get_ufuncimpl(PyUFuncObject *ufunc,
     }
 
     if (force_legacy_promotion
-            && npy_promotion_state == NPY_USE_LEGACY_PROMOTION) {
+            && npy_promotion_state == NPY_USE_LEGACY_PROMOTION
+            && (ufunc->ntypes != 0 || ufunc->userloops != NULL)) {
         /*
          * We must use legacy promotion for value-based logic. Call the old
          * resolver once up-front to get the "actual" loop dtypes.

--- a/numpy/_core/src/umath/string_buffer.h
+++ b/numpy/_core/src/umath/string_buffer.h
@@ -1,0 +1,209 @@
+#ifndef _NPY_CORE_SRC_UMATH_STRING_BUFFER_H_
+#define _NPY_CORE_SRC_UMATH_STRING_BUFFER_H_
+
+#include <Python.h>
+#include <cstddef>
+
+#define NPY_NO_DEPRECATED_API NPY_API_VERSION
+#define _MULTIARRAYMODULE
+#define _UMATHMODULE
+
+#include "numpy/ndarraytypes.h"
+
+#define CHECK_OVERFLOW(index) if (buf + (index) >= after) return 0
+#define MSB(val) ((val) >> 7 & 1)
+
+
+enum class ENCODING {
+    ASCII, UTF32
+};
+
+
+template <ENCODING enc>
+inline npy_ucs4
+getchar(const unsigned char *buf, int *bytes);
+
+
+template <>
+inline npy_ucs4
+getchar<ENCODING::ASCII>(const unsigned char *buf, int *bytes)
+{
+    *bytes = 1;
+    return (npy_ucs4) *buf;
+}
+
+
+template <>
+inline npy_ucs4
+getchar<ENCODING::UTF32>(const unsigned char *buf, int *bytes)
+{
+    *bytes = 4;
+    return *(npy_ucs4 *)buf;
+}
+
+
+template <ENCODING enc>
+struct Buffer {
+    char *buf;
+    char *after;
+
+    inline Buffer<enc>() {
+        buf = after = NULL;
+    }
+
+    inline Buffer<enc>(char *buf_, int elsize_) {
+        buf = buf_;
+        after = buf_ + elsize_;
+    }
+
+    inline npy_int64 length() {
+        Buffer tmp(after, 0);
+        tmp--;
+        while (tmp >= *this && *tmp == '\0') {
+            tmp--;
+        }
+        return (npy_int64) (tmp - *this + 1);
+    }
+
+    inline Buffer<enc>& operator+=(npy_int64 rhs) {
+        switch (enc) {
+        case ENCODING::ASCII:
+            buf += rhs;
+            break;
+        case ENCODING::UTF32:
+            buf += rhs * sizeof(npy_ucs4);
+            break;
+        }
+        return *this;
+    }
+
+    inline Buffer<enc>& operator-=(npy_int64 rhs) {
+        switch (enc) {
+        case ENCODING::ASCII:
+            buf -= rhs;
+            break;
+        case ENCODING::UTF32:
+            buf -= rhs * sizeof(npy_ucs4);
+            break;
+        }
+        return *this;
+    }
+
+    inline Buffer<enc>& operator++() {
+        *this += 1;
+        return *this;
+    }
+
+    inline Buffer<enc> operator++(int) {
+        Buffer<enc> old = *this;
+        operator++();
+        return old; 
+    }
+
+    inline Buffer<enc>& operator--() {
+        *this -= 1;
+        return *this;
+    }
+
+    inline Buffer<enc> operator--(int) {
+        Buffer<enc> old = *this;
+        operator--();
+        return old; 
+    }
+
+    inline npy_ucs4 operator*() {
+        int bytes;
+        return getchar<enc>((unsigned char *) buf, &bytes);
+    }
+
+    inline npy_ucs4 operator[](size_t index) {
+        int bytes;
+        switch (enc) {
+        case ENCODING::ASCII:
+            CHECK_OVERFLOW(index);
+            return getchar<enc>((unsigned char *) (buf + index), &bytes);
+        case ENCODING::UTF32:
+            CHECK_OVERFLOW(index * sizeof(npy_ucs4));
+            return getchar<enc>((unsigned char *) (buf + index * sizeof(npy_ucs4)), &bytes);
+        }
+    }
+
+    inline Buffer<enc> fast_memchr(npy_ucs4 ch, int len) {
+        switch (enc) {
+        case ENCODING::ASCII:
+            buf = (char *) memchr(buf, ch, len);
+            return *this;
+        case ENCODING::UTF32:
+            buf = (char *) wmemchr((wchar_t *) buf, ch, len);
+            return *this;
+        }
+    }
+
+    inline int fast_memcmp(Buffer<enc> other, int len) {
+        return memcmp(buf, other.buf, len);
+    }
+};
+
+
+template <ENCODING enc>
+inline Buffer<enc> operator+(Buffer<enc> lhs, npy_int64 rhs) {
+    lhs += rhs;
+    return lhs;
+}
+
+
+template <ENCODING enc>
+inline std::ptrdiff_t operator-(Buffer<enc> lhs, Buffer<enc> rhs) {
+    switch (enc) {
+    case ENCODING::ASCII:
+        return lhs.buf - rhs.buf;
+    case ENCODING::UTF32:
+        return (lhs.buf - rhs.buf) / (std::ptrdiff_t) sizeof(npy_ucs4);
+    }
+}
+
+
+template <ENCODING enc>
+inline Buffer<enc> operator-(Buffer<enc> lhs, npy_int64 rhs) {
+    lhs -= rhs;
+    return lhs;
+}
+
+
+template <ENCODING enc>
+inline bool operator==(Buffer<enc> lhs, Buffer<enc> rhs) {
+    return lhs.buf == rhs.buf;
+}
+
+
+template <ENCODING enc>
+inline bool operator!=(Buffer<enc> lhs, Buffer<enc> rhs) {
+    return !(rhs == lhs);
+}
+
+
+template <ENCODING enc>
+inline bool operator<(Buffer<enc> lhs, Buffer<enc> rhs) {
+    return lhs.buf < rhs.buf;
+}
+
+
+template <ENCODING enc>
+inline bool operator>(Buffer<enc> lhs, Buffer<enc> rhs) {
+    return rhs < lhs;
+}
+
+
+template <ENCODING enc>
+inline bool operator<=(Buffer<enc> lhs, Buffer<enc> rhs) {
+    return !(lhs > rhs);
+}
+
+
+template <ENCODING enc>
+inline bool operator>=(Buffer<enc> lhs, Buffer<enc> rhs) {
+    return !(lhs < rhs);
+}
+
+
+#endif /* _NPY_CORE_SRC_UMATH_STRING_BUFFER_H_ */

--- a/numpy/_core/src/umath/string_fastsearch.h
+++ b/numpy/_core/src/umath/string_fastsearch.h
@@ -1,0 +1,724 @@
+#ifndef _NPY_CORE_SRC_UMATH_STRING_FASTSEARCH_H_
+#define _NPY_CORE_SRC_UMATH_STRING_FASTSEARCH_H_
+
+/* stringlib: fastsearch implementation taken from CPython */
+
+#include <Python.h>
+#include <limits.h>
+#include <string.h>
+#include <wchar.h>
+
+#include <type_traits>
+
+#include "string_buffer.h"
+
+
+/* fast search/count implementation, based on a mix between boyer-
+   moore and horspool, with a few more bells and whistles on the top.
+   for some more background, see:
+   https://web.archive.org/web/20201107074620/http://effbot.org/zone/stringlib.htm */
+
+/* note: fastsearch may access s[n], but this is being checked for in this
+   implementation, because NumPy strings are not NULL-terminated.
+   also, the count mode returns -1 if there cannot possibly be a match
+   in the target string, and 0 if it has actually checked for matches,
+   but didn't find any. callers beware! */
+
+/* If the strings are long enough, use Crochemore and Perrin's Two-Way
+   algorithm, which has worst-case O(n) runtime and best-case O(n/k).
+   Also compute a table of shifts to achieve O(n/k) in more cases,
+   and often (data dependent) deduce larger shifts than pure C&P can
+   deduce. See stringlib_find_two_way_notes.txt in this folder for a
+   detailed explanation. */
+
+#define FAST_COUNT 0
+#define FAST_SEARCH 1
+#define FAST_RSEARCH 2
+
+#if LONG_BIT >= 128
+#define STRINGLIB_BLOOM_WIDTH 128
+#elif LONG_BIT >= 64
+#define STRINGLIB_BLOOM_WIDTH 64
+#elif LONG_BIT >= 32
+#define STRINGLIB_BLOOM_WIDTH 32
+#else
+#error "LONG_BIT is smaller than 32"
+#endif
+
+#define STRINGLIB_BLOOM_ADD(mask, ch) \
+    ((mask |= (1UL << ((ch) & (STRINGLIB_BLOOM_WIDTH -1)))))
+#define STRINGLIB_BLOOM(mask, ch)     \
+    ((mask &  (1UL << ((ch) & (STRINGLIB_BLOOM_WIDTH -1)))))
+
+#define FORWARD_DIRECTION 1
+#define BACKWARD_DIRECTION -1
+#define MEMCHR_CUT_OFF 15
+
+
+template <ENCODING enc>
+inline Py_ssize_t
+findchar(Buffer<enc> s, Py_ssize_t n, npy_ucs4 ch)
+{
+    Buffer<enc> p = s, e = s + n;
+
+    if (n > MEMCHR_CUT_OFF) {
+        p = s.fast_memchr(ch, n);
+        if (p.buf != NULL) {
+            return (p - s);
+        }
+        return -1;
+    }
+    while (p < e) {
+        if (*p == ch) {
+            return (p - s);
+        }
+        p++;
+    }
+    return -1;
+}
+
+template <ENCODING enc>
+inline Py_ssize_t
+rfindchar(Buffer<enc> s, Py_ssize_t n, npy_ucs4 ch)
+{
+    Buffer p = s + n;
+    while (p > s) {
+        p--;
+        if (*p == ch)
+            return (p - s);
+    }
+    return -1;
+}
+
+
+/* Change to a 1 to see logging comments walk through the algorithm. */
+#if 0 && STRINGLIB_SIZEOF_CHAR == 1
+# define LOG(...) printf(__VA_ARGS__)
+# define LOG_STRING(s, n) printf("\"%.*s\"", (int)(n), s)
+# define LOG_LINEUP() do {                                         \
+    LOG("> "); LOG_STRING(haystack, len_haystack); LOG("\n> ");    \
+    LOG("%*s",(int)(window_last - haystack + 1 - len_needle), ""); \
+    LOG_STRING(needle, len_needle); LOG("\n");                     \
+} while(0)
+#else
+# define LOG(...)
+# define LOG_STRING(s, n)
+# define LOG_LINEUP()
+#endif
+
+template <ENCODING enc>
+static inline Py_ssize_t
+_lex_search(Buffer<enc> needle, Py_ssize_t len_needle,
+            Py_ssize_t *return_period, int invert_alphabet)
+{
+    /* Do a lexicographic search. Essentially this:
+           >>> max(needle[i:] for i in range(len(needle)+1))
+       Also find the period of the right half.   */
+    Py_ssize_t max_suffix = 0;
+    Py_ssize_t candidate = 1;
+    Py_ssize_t k = 0;
+    // The period of the right half.
+    Py_ssize_t period = 1;
+
+    while (candidate + k < len_needle) {
+        // each loop increases candidate + k + max_suffix
+        npy_ucs4 a = needle[candidate + k];
+        npy_ucs4 b = needle[max_suffix + k];
+        // check if the suffix at candidate is better than max_suffix
+        if (invert_alphabet ? (b < a) : (a < b)) {
+            // Fell short of max_suffix.
+            // The next k + 1 characters are non-increasing
+            // from candidate, so they won't start a maximal suffix.
+            candidate += k + 1;
+            k = 0;
+            // We've ruled out any period smaller than what's
+            // been scanned since max_suffix.
+            period = candidate - max_suffix;
+        }
+        else if (a == b) {
+            if (k + 1 != period) {
+                // Keep scanning the equal strings
+                k++;
+            }
+            else {
+                // Matched a whole period.
+                // Start matching the next period.
+                candidate += period;
+                k = 0;
+            }
+        }
+        else {
+            // Did better than max_suffix, so replace it.
+            max_suffix = candidate;
+            candidate++;
+            k = 0;
+            period = 1;
+        }
+    }
+    *return_period = period;
+    return max_suffix;
+}
+
+template <ENCODING enc>
+static inline Py_ssize_t
+_factorize(Buffer<enc> needle,
+                      Py_ssize_t len_needle,
+                      Py_ssize_t *return_period)
+{
+    /* Do a "critical factorization", making it so that:
+       >>> needle = (left := needle[:cut]) + (right := needle[cut:])
+       where the "local period" of the cut is maximal.
+
+       The local period of the cut is the minimal length of a string w
+       such that (left endswith w or w endswith left)
+       and (right startswith w or w startswith left).
+
+       The Critical Factorization Theorem says that this maximal local
+       period is the global period of the string.
+
+       Crochemore and Perrin (1991) show that this cut can be computed
+       as the later of two cuts: one that gives a lexicographically
+       maximal right half, and one that gives the same with the
+       with respect to a reversed alphabet-ordering.
+
+       This is what we want to happen:
+           >>> x = "GCAGAGAG"
+           >>> cut, period = factorize(x)
+           >>> x[:cut], (right := x[cut:])
+           ('GC', 'AGAGAG')
+           >>> period  # right half period
+           2
+           >>> right[period:] == right[:-period]
+           True
+
+       This is how the local period lines up in the above example:
+                GC | AGAGAG
+           AGAGAGC = AGAGAGC
+       The length of this minimal repetition is 7, which is indeed the
+       period of the original string. */
+
+    Py_ssize_t cut1, period1, cut2, period2, cut, period;
+    cut1 = _lex_search<enc>(needle, len_needle, &period1, 0);
+    cut2 = _lex_search<enc>(needle, len_needle, &period2, 1);
+
+    // Take the later cut.
+    if (cut1 > cut2) {
+        period = period1;
+        cut = cut1;
+    }
+    else {
+        period = period2;
+        cut = cut2;
+    }
+
+    LOG("split: "); LOG_STRING(needle, cut);
+    LOG(" + "); LOG_STRING(needle + cut, len_needle - cut);
+    LOG("\n");
+
+    *return_period = period;
+    return cut;
+}
+
+
+#define SHIFT_TYPE uint8_t
+#define MAX_SHIFT UINT8_MAX
+
+#define TABLE_SIZE_BITS 6u
+#define TABLE_SIZE (1U << TABLE_SIZE_BITS)
+#define TABLE_MASK (TABLE_SIZE - 1U)
+
+template <ENCODING enc>
+struct prework {
+    Buffer<enc> needle;
+    Py_ssize_t len_needle;
+    Py_ssize_t cut;
+    Py_ssize_t period;
+    Py_ssize_t gap;
+    int is_periodic;
+    SHIFT_TYPE table[TABLE_SIZE];
+};
+
+
+template <ENCODING enc>
+static void
+_preprocess(Buffer<enc> needle, Py_ssize_t len_needle,
+            prework<enc> *p)
+{
+    p->needle = needle;
+    p->len_needle = len_needle;
+    p->cut = _factorize(needle, len_needle, &(p->period));
+    assert(p->period + p->cut <= len_needle);
+    p->is_periodic = (0 == needle.fast_memcmp(needle + p->period,
+                                              p->cut * sizeof(npy_ucs4)));
+    if (p->is_periodic) {
+        assert(p->cut <= len_needle/2);
+        assert(p->cut < p->period);
+        p->gap = 0; // unused
+    }
+    else {
+        // A lower bound on the period
+        p->period = Py_MAX(p->cut, len_needle - p->cut) + 1;
+        // The gap between the last character and the previous
+        // occurrence of an equivalent character (modulo TABLE_SIZE)
+        p->gap = len_needle;
+        npy_ucs4 last = needle[len_needle - 1] & TABLE_MASK;
+        for (Py_ssize_t i = len_needle - 2; i >= 0; i--) {
+            npy_ucs4 x = needle[i] & TABLE_MASK;
+            if (x == last) {
+                p->gap = len_needle - 1 - i;
+                break;
+            }
+        }
+    }
+    // Fill up a compressed Boyer-Moore "Bad Character" table
+    Py_ssize_t not_found_shift = Py_MIN(len_needle, MAX_SHIFT);
+    for (Py_ssize_t i = 0; i < (Py_ssize_t)TABLE_SIZE; i++) {
+        p->table[i] = Py_SAFE_DOWNCAST(not_found_shift,
+                                       Py_ssize_t, SHIFT_TYPE);
+    }
+    for (Py_ssize_t i = len_needle - not_found_shift; i < len_needle; i++) {
+        SHIFT_TYPE shift = Py_SAFE_DOWNCAST(len_needle - 1 - i,
+                                            Py_ssize_t, SHIFT_TYPE);
+        p->table[needle[i] & TABLE_MASK] = shift;
+    }
+}
+
+template <ENCODING enc>
+static Py_ssize_t
+_two_way(Buffer<enc> haystack, Py_ssize_t len_haystack,
+         prework<enc> *p)
+{
+    // Crochemore and Perrin's (1991) Two-Way algorithm.
+    // See http://www-igm.univ-mlv.fr/~lecroq/string/node26.html#SECTION00260
+    const Py_ssize_t len_needle = p->len_needle;
+    const Py_ssize_t cut = p->cut;
+    Py_ssize_t period = p->period;
+    Buffer<enc> needle = p->needle;
+    Buffer<enc> window_last = haystack + len_needle - 1;
+    Buffer<enc> haystack_end = haystack + len_haystack;
+    SHIFT_TYPE *table = p->table;
+    Buffer<enc> window;
+    LOG("===== Two-way: \"%s\" in \"%s\". =====\n", needle, haystack);
+
+    if (p->is_periodic) {
+        LOG("Needle is periodic.\n");
+        Py_ssize_t memory = 0;
+      periodicwindowloop:
+        while (window_last < haystack_end) {
+            assert(memory == 0);
+            for (;;) {
+                LOG_LINEUP();
+                Py_ssize_t shift = table[window_last[0] & TABLE_MASK];
+                window_last += shift;
+                if (shift == 0) {
+                    break;
+                }
+                if (window_last >= haystack_end) {
+                    return -1;
+                }
+                LOG("Horspool skip\n");
+            }
+          no_shift:
+            window = window_last - len_needle + 1;
+            assert((window[len_needle - 1] & TABLE_MASK) ==
+                   (needle[len_needle - 1] & TABLE_MASK));
+            Py_ssize_t i = Py_MAX(cut, memory);
+            for (; i < len_needle; i++) {
+                if (needle[i] != window[i]) {
+                    LOG("Right half does not match.\n");
+                    window_last += i - cut + 1;
+                    memory = 0;
+                    goto periodicwindowloop;
+                }
+            }
+            for (i = memory; i < cut; i++) {
+                if (needle[i] != window[i]) {
+                    LOG("Left half does not match.\n");
+                    window_last += period;
+                    memory = len_needle - period;
+                    if (window_last >= haystack_end) {
+                        return -1;
+                    }
+                    Py_ssize_t shift = table[window_last[0] & TABLE_MASK];
+                    if (shift) {
+                        // A mismatch has been identified to the right
+                        // of where i will next start, so we can jump
+                        // at least as far as if the mismatch occurred
+                        // on the first comparison.
+                        Py_ssize_t mem_jump = Py_MAX(cut, memory) - cut + 1;
+                        LOG("Skip with Memory.\n");
+                        memory = 0;
+                        window_last += Py_MAX(shift, mem_jump);
+                        goto periodicwindowloop;
+                    }
+                    goto no_shift;
+                }
+            }
+            LOG("Found a match!\n");
+            return window - haystack;
+        }
+    }
+    else {
+        Py_ssize_t gap = p->gap;
+        period = Py_MAX(gap, period);
+        LOG("Needle is not periodic.\n");
+        Py_ssize_t gap_jump_end = Py_MIN(len_needle, cut + gap);
+      windowloop:
+        while (window_last < haystack_end) {
+            for (;;) {
+                LOG_LINEUP();
+                Py_ssize_t shift = table[window_last[0] & TABLE_MASK];
+                window_last += shift;
+                if (shift == 0) {
+                    break;
+                }
+                if (window_last >= haystack_end) {
+                    return -1;
+                }
+                LOG("Horspool skip\n");
+            }
+            window = window_last - len_needle + 1;
+            assert((window[len_needle - 1] & TABLE_MASK) ==
+                   (needle[len_needle - 1] & TABLE_MASK));
+            for (Py_ssize_t i = cut; i < gap_jump_end; i++) {
+                if (needle[i] != window[i]) {
+                    LOG("Early right half mismatch: jump by gap.\n");
+                    assert(gap >= i - cut + 1);
+                    window_last += gap;
+                    goto windowloop;
+                }
+            }
+            for (Py_ssize_t i = gap_jump_end; i < len_needle; i++) {
+                if (needle[i] != window[i]) {
+                    LOG("Late right half mismatch.\n");
+                    assert(i - cut + 1 > gap);
+                    window_last += i - cut + 1;
+                    goto windowloop;
+                }
+            }
+            for (Py_ssize_t i = 0; i < cut; i++) {
+                if (needle[i] != window[i]) {
+                    LOG("Left half does not match.\n");
+                    window_last += period;
+                    goto windowloop;
+                }
+            }
+            LOG("Found a match!\n");
+            return window - haystack;
+        }
+    }
+    LOG("Not found. Returning -1.\n");
+    return -1;
+}
+
+
+template <ENCODING enc>
+static inline Py_ssize_t
+_two_way_find(Buffer<enc> haystack, Py_ssize_t len_haystack,
+              Buffer<enc> needle, Py_ssize_t len_needle)
+{
+    LOG("###### Finding \"%s\" in \"%s\".\n", needle, haystack);
+    prework<enc> p;
+    _preprocess(needle, len_needle, &p);
+    return _two_way(haystack, len_haystack, &p);
+}
+
+
+template <ENCODING enc>
+static inline Py_ssize_t
+_two_way_count(Buffer<enc> haystack, Py_ssize_t len_haystack,
+               Buffer<enc> needle, Py_ssize_t len_needle,
+               Py_ssize_t maxcount)
+{
+    LOG("###### Counting \"%s\" in \"%s\".\n", needle, haystack);
+    prework<enc> p;
+    _preprocess(needle, len_needle, &p);
+    Py_ssize_t index = 0, count = 0;
+    while (1) {
+        Py_ssize_t result;
+        result = _two_way(haystack + index,
+                                     len_haystack - index, &p);
+        if (result == -1) {
+            return count;
+        }
+        count++;
+        if (count == maxcount) {
+            return maxcount;
+        }
+        index += result + len_needle;
+    }
+    return count;
+}
+
+#undef SHIFT_TYPE
+#undef NOT_FOUND
+#undef SHIFT_OVERFLOW
+#undef TABLE_SIZE_BITS
+#undef TABLE_SIZE
+#undef TABLE_MASK
+
+#undef LOG
+#undef LOG_STRING
+#undef LOG_LINEUP
+
+template <ENCODING enc>
+static inline Py_ssize_t
+default_find(Buffer<enc> s, Py_ssize_t n,
+             Buffer<enc> p, Py_ssize_t m,
+             Py_ssize_t maxcount, int mode)
+{
+    const Py_ssize_t w = n - m;
+    Py_ssize_t mlast = m - 1, count = 0;
+    Py_ssize_t gap = mlast;
+    const npy_ucs4 last = p[mlast];
+    Buffer<enc> ss = s + mlast;
+
+    unsigned long mask = 0;
+    for (Py_ssize_t i = 0; i < mlast; i++) {
+        STRINGLIB_BLOOM_ADD(mask, p[i]);
+        if (p[i] == last) {
+            gap = mlast - i - 1;
+        }
+    }
+    STRINGLIB_BLOOM_ADD(mask, last);
+
+    for (Py_ssize_t i = 0; i <= w; i++) {
+        if (ss[i] == last) {
+            /* candidate match */
+            Py_ssize_t j;
+            for (j = 0; j < mlast; j++) {
+                if (s[i+j] != p[j]) {
+                    break;
+                }
+            }
+            if (j == mlast) {
+                /* got a match! */
+                if (mode != FAST_COUNT) {
+                    return i;
+                }
+                count++;
+                if (count == maxcount) {
+                    return maxcount;
+                }
+                i = i + mlast;
+                continue;
+            }
+            /* miss: check if next character is part of pattern */
+            if (!STRINGLIB_BLOOM(mask, ss[i+1])) {
+                i = i + m;
+            }
+            else {
+                i = i + gap;
+            }
+        }
+        else {
+            /* skip: check if next character is part of pattern */
+            if (!STRINGLIB_BLOOM(mask, ss[i+1])) {
+                i = i + m;
+            }
+        }
+    }
+    return mode == FAST_COUNT ? count : -1;
+}
+
+
+template <ENCODING enc>
+static Py_ssize_t
+adaptive_find(Buffer<enc> s, Py_ssize_t n,
+              Buffer<enc> p, Py_ssize_t m,
+              Py_ssize_t maxcount, int mode)
+{
+    const Py_ssize_t w = n - m;
+    Py_ssize_t mlast = m - 1, count = 0;
+    Py_ssize_t gap = mlast;
+    Py_ssize_t hits = 0, res;
+    const npy_ucs4 last = p[mlast];
+    Buffer<enc> ss = s + mlast;
+
+    unsigned long mask = 0;
+    for (Py_ssize_t i = 0; i < mlast; i++) {
+        STRINGLIB_BLOOM_ADD(mask, p[i]);
+        if (p[i] == last) {
+            gap = mlast - i - 1;
+        }
+    }
+    STRINGLIB_BLOOM_ADD(mask, last);
+
+    for (Py_ssize_t i = 0; i <= w; i++) {
+        if (ss[i] == last) {
+            /* candidate match */
+            Py_ssize_t j;
+            for (j = 0; j < mlast; j++) {
+                if (s[i+j] != p[j]) {
+                    break;
+                }
+            }
+            if (j == mlast) {
+                /* got a match! */
+                if (mode != FAST_COUNT) {
+                    return i;
+                }
+                count++;
+                if (count == maxcount) {
+                    return maxcount;
+                }
+                i = i + mlast;
+                continue;
+            }
+            hits += j + 1;
+            if (hits > m / 4 && w - i > 2000) {
+                if (mode == FAST_SEARCH) {
+                    res = _two_way_find(s + i, n - i, p, m);
+                    return res == -1 ? -1 : res + i;
+                }
+                else {
+                    res = _two_way_count(s + i, n - i, p, m, maxcount - count);
+                    return res + count;
+                }
+            }
+            /* miss: check if next character is part of pattern */
+            if (!STRINGLIB_BLOOM(mask, ss[i+1])) {
+                i = i + m;
+            }
+            else {
+                i = i + gap;
+            }
+        }
+        else {
+            /* skip: check if next character is part of pattern */
+            if (!STRINGLIB_BLOOM(mask, ss[i+1])) {
+                i = i + m;
+            }
+        }
+    }
+    return mode == FAST_COUNT ? count : -1;
+}
+
+
+template <ENCODING enc>
+static Py_ssize_t
+default_rfind(Buffer<enc> s, Py_ssize_t n,
+              Buffer<enc> p, Py_ssize_t m,
+              Py_ssize_t maxcount, int mode)
+{
+    /* create compressed boyer-moore delta 1 table */
+    unsigned long mask = 0;
+    Py_ssize_t i, j, mlast = m - 1, skip = m - 1, w = n - m;
+
+    /* process pattern[0] outside the loop */
+    STRINGLIB_BLOOM_ADD(mask, p[0]);
+    /* process pattern[:0:-1] */
+    for (i = mlast; i > 0; i--) {
+        STRINGLIB_BLOOM_ADD(mask, p[i]);
+        if (p[i] == p[0]) {
+            skip = i - 1;
+        }
+    }
+
+    for (i = w; i >= 0; i--) {
+        if (s[i] == p[0]) {
+            /* candidate match */
+            for (j = mlast; j > 0; j--) {
+                if (s[i+j] != p[j]) {
+                    break;
+                }
+            }
+            if (j == 0) {
+                /* got a match! */
+                return i;
+            }
+            /* miss: check if previous character is part of pattern */
+            if (i > 0 && !STRINGLIB_BLOOM(mask, s[i-1])) {
+                i = i - m;
+            }
+            else {
+                i = i - skip;
+            }
+        }
+        else {
+            /* skip: check if previous character is part of pattern */
+            if (i > 0 && !STRINGLIB_BLOOM(mask, s[i-1])) {
+                i = i - m;
+            }
+        }
+    }
+    return -1;
+}
+
+
+template <ENCODING enc>
+static inline Py_ssize_t
+countchar(Buffer<enc> s, Py_ssize_t n,
+          const npy_ucs4 p0, Py_ssize_t maxcount)
+{
+    Py_ssize_t i, count = 0;
+    for (i = 0; i < n; i++) {
+        if (s[i] == p0) {
+            count++;
+            if (count == maxcount) {
+                return maxcount;
+            }
+        }
+    }
+    return count;
+}
+
+
+template <ENCODING enc>
+inline Py_ssize_t
+fastsearch(Buffer<enc> s, Py_ssize_t n,
+           Buffer<enc> p, Py_ssize_t m,
+           Py_ssize_t maxcount, int mode)
+{
+    if (n < m || (mode == FAST_COUNT && maxcount == 0)) {
+        return -1;
+    }
+
+    /* look for special cases */
+    if (m <= 1) {
+        if (m <= 0) {
+            return -1;
+        }
+        /* use special case for 1-character strings */
+        if (mode == FAST_SEARCH)
+            return findchar(s, n, p[0]);
+        else if (mode == FAST_RSEARCH)
+            return rfindchar(s, n, p[0]);
+        else {
+            return countchar(s, n, p[0], maxcount);
+        }
+    }
+
+    if (mode != FAST_RSEARCH) {
+        if (n < 2500 || (m < 100 && n < 30000) || m < 6) {
+            return default_find(s, n, p, m, maxcount, mode);
+        }
+        else if ((m >> 2) * 3 < (n >> 2)) {
+            /* 33% threshold, but don't overflow. */
+            /* For larger problems where the needle isn't a huge
+               percentage of the size of the haystack, the relatively
+               expensive O(m) startup cost of the two-way algorithm
+               will surely pay off. */
+            if (mode == FAST_SEARCH) {
+                return _two_way_find(s, n, p, m);
+            }
+            else {
+                return _two_way_count(s, n, p, m, maxcount);
+            }
+        }
+        else {
+            /* To ensure that we have good worst-case behavior,
+               here's an adaptive version of the algorithm, where if
+               we match O(m) characters without any matches of the
+               entire needle, then we predict that the startup cost of
+               the two-way algorithm will probably be worth it. */
+            return adaptive_find(s, n, p, m, maxcount, mode);
+        }
+    }
+    else {
+        /* FAST_RSEARCH */
+        return default_rfind(s, n, p, m, maxcount, mode);
+    }
+}
+
+#endif  /* _NPY_CORE_SRC_UMATH_STRING_FASTSEARCH_H_ */

--- a/numpy/_core/src/umath/string_ufuncs.cpp
+++ b/numpy/_core/src/umath/string_ufuncs.cpp
@@ -228,7 +228,7 @@ adjust_offsets(npy_int64 *start, npy_int64 *end, npy_int64 len)
 }
 
 template <ENCODING enc>
-static inline npy_long
+static inline npy_intp
 string_find(char *str1, int elsize1, char *str2, int elsize2,
             npy_int64 start, npy_int64 end)
 {
@@ -240,20 +240,20 @@ string_find(char *str1, int elsize1, char *str2, int elsize2,
 
     adjust_offsets(&start, &end, len1);
     if (end - start < len2) {
-        return (npy_long) -1;
+        return (npy_intp) -1;
     }
 
     if (len2 == 1) {
         npy_ucs4 ch = *buf2;
-        npy_long result = (npy_long) findchar<enc>(buf1 + start, end - start, ch);
+        npy_intp result = (npy_intp) findchar<enc>(buf1 + start, end - start, ch);
         if (result == -1) {
-            return (npy_long) -1;
+            return (npy_intp) -1;
         } else {
-            return result + (npy_long) start;
+            return result + (npy_intp) start;
         }
     }
 
-    return (npy_long) findslice<enc>(buf1 + start,
+    return (npy_intp) findslice<enc>(buf1 + start,
                                      end - start,
                                      buf2,
                                      len2,
@@ -261,7 +261,7 @@ string_find(char *str1, int elsize1, char *str2, int elsize2,
 }
 
 template <ENCODING enc>
-static inline npy_long
+static inline npy_intp
 string_rfind(char *str1, int elsize1, char *str2, int elsize2,
              npy_int64 start, npy_int64 end)
 {
@@ -273,20 +273,20 @@ string_rfind(char *str1, int elsize1, char *str2, int elsize2,
 
     adjust_offsets(&start, &end, len1);
     if (end - start < len2) {
-        return (npy_long) -1;
+        return (npy_intp) -1;
     }
 
     if (len2 == 1) {
         npy_ucs4 ch = *buf2;
-        npy_long result = (npy_long) rfindchar(buf1 + start, end - start, ch);
+        npy_intp result = (npy_intp) rfindchar(buf1 + start, end - start, ch);
         if (result == -1) {
-            return (npy_long) -1;
+            return (npy_intp) -1;
         } else {
-            return result + (npy_long) start;
+            return result + (npy_intp) start;
         }
     }
 
-    return (npy_long) rfindslice(buf1 + start,
+    return (npy_intp) rfindslice(buf1 + start,
                                  end - start,
                                  buf2,
                                  len2,
@@ -448,9 +448,9 @@ string_find_loop(PyArrayMethod_Context *context,
     npy_intp N = dimensions[0];
 
     while (N--) {
-        npy_long idx = string_find<enc>(in1, elsize1, in2, elsize2,
+        npy_intp idx = string_find<enc>(in1, elsize1, in2, elsize2,
                                          *(npy_int64 *)in3, *(npy_int64 *)in4);
-        *(npy_long *)out = idx;
+        *(npy_intp *)out = idx;
 
         in1 += strides[0];
         in2 += strides[1];
@@ -480,9 +480,9 @@ string_rfind_loop(PyArrayMethod_Context *context,
     npy_intp N = dimensions[0];
 
     while (N--) {
-        npy_long idx = string_rfind<enc>(in1, elsize1, in2, elsize2,
+        npy_intp idx = string_rfind<enc>(in1, elsize1, in2, elsize2,
                                           *(npy_int64 *)in3, *(npy_int64 *)in4);
-        *(npy_long *)out = idx;
+        *(npy_intp *)out = idx;
 
         in1 += strides[0];
         in2 += strides[1];
@@ -534,7 +534,7 @@ string_find_rfind_promoter(PyUFuncObject *NPY_UNUSED(ufunc),
     new_op_dtypes[1] = op_dtypes[1];
     new_op_dtypes[2] = PyArray_DTypeFromTypeNum(NPY_INT64);
     new_op_dtypes[3] = PyArray_DTypeFromTypeNum(NPY_INT64);
-    new_op_dtypes[4] = PyArray_DTypeFromTypeNum(NPY_LONG);
+    new_op_dtypes[4] = PyArray_DTypeFromTypeNum(NPY_DEFAULT_INT);
     return 0;
 }
 
@@ -779,7 +779,7 @@ init_string_ufuncs(PyObject *umath)
 
     dtypes[0] = dtypes[1] = NPY_STRING;
     dtypes[2] = dtypes[3] = NPY_INT64;
-    dtypes[4] = NPY_LONG;
+    dtypes[4] = NPY_DEFAULT_INT;
     if (init_ufunc(umath, "find", "templated_string_find", 4, 1, dtypes,
                    string_find_loop<ENCODING::ASCII>, string_find_loop<ENCODING::UTF32>,
                    NULL, string_find_rfind_promoter) < 0) {
@@ -788,7 +788,7 @@ init_string_ufuncs(PyObject *umath)
 
     dtypes[0] = dtypes[1] = NPY_STRING;
     dtypes[2] = dtypes[3] = NPY_INT64;
-    dtypes[4] = NPY_LONG;
+    dtypes[4] = NPY_DEFAULT_INT;
     if (init_ufunc(umath, "rfind", "templated_string_rfind", 4, 1, dtypes,
                    string_rfind_loop<ENCODING::ASCII>, string_rfind_loop<ENCODING::UTF32>,
                    NULL, string_find_rfind_promoter) < 0) {

--- a/numpy/_core/src/umath/string_ufuncs.cpp
+++ b/numpy/_core/src/umath/string_ufuncs.cpp
@@ -16,6 +16,8 @@
 #include "convert_datatype.h"
 
 #include "string_ufuncs.h"
+#include "string_fastsearch.h"
+#include "string_buffer.h"
 
 
 template <typename character>
@@ -64,6 +66,51 @@ get_length(const character *str, int elsize)
     return d - str + 1;
 }
 
+
+template <ENCODING enc>
+static inline Py_ssize_t
+findslice(Buffer<enc> str, Py_ssize_t str_len,
+          Buffer<enc> sub, Py_ssize_t sub_len,
+          Py_ssize_t offset)
+{
+    Py_ssize_t pos;
+
+    if (sub_len == 0) {
+        return offset;
+    }
+
+    pos = fastsearch<enc>(str, str_len, sub, sub_len, -1, FAST_SEARCH);
+
+    if (pos >= 0) {
+        pos += offset;
+    }
+
+    return pos;
+}
+
+
+template <ENCODING enc>
+static inline Py_ssize_t
+rfindslice(Buffer<enc> str, Py_ssize_t str_len,
+           Buffer<enc> sub, Py_ssize_t sub_len,
+           Py_ssize_t offset)
+{
+    Py_ssize_t pos;
+
+    if (sub_len == 0) {
+        // Note: This won't work for utf-8 and should be fixed before adding
+        // support for stringdtype.
+        return str_len + offset;
+    }
+
+    pos = fastsearch<enc>(str, str_len, sub, sub_len, -1, FAST_RSEARCH);
+
+    if (pos >= 0) {
+        pos += offset;
+    }
+
+    return pos;
+}
 
 /*
  * Compare two strings of different length.  Note that either string may be
@@ -157,6 +204,93 @@ string_isalpha(const character *str, int elsize)
         str++;
     }
     return (npy_bool) 1;
+}
+
+/* helper to fixup start/end slice values */
+static inline void
+adjust_offsets(npy_int64 *start, npy_int64 *end, npy_int64 len)
+{
+    if (*end > len) {
+        *end = len;
+    } else if (*end < 0) {
+        *end += len;
+        if (*end < 0) {
+            *end = 0;
+        }
+    }
+
+    if (*start < 0) {
+        *start += len;
+        if (*start < 0) {
+            *start = 0;
+        }
+    }
+}
+
+template <ENCODING enc>
+static inline npy_long
+string_find(char *str1, int elsize1, char *str2, int elsize2,
+            npy_int64 start, npy_int64 end)
+{
+    Buffer<enc> buf1(str1, elsize1);
+    Buffer<enc> buf2(str2, elsize2);
+
+    npy_int64 len1 = buf1.length();
+    npy_int64 len2 = buf2.length();
+
+    adjust_offsets(&start, &end, len1);
+    if (end - start < len2) {
+        return (npy_long) -1;
+    }
+
+    if (len2 == 1) {
+        npy_ucs4 ch = *buf2;
+        npy_long result = (npy_long) findchar<enc>(buf1 + start, end - start, ch);
+        if (result == -1) {
+            return (npy_long) -1;
+        } else {
+            return result + (npy_long) start;
+        }
+    }
+
+    return (npy_long) findslice<enc>(buf1 + start,
+                                     end - start,
+                                     buf2,
+                                     len2,
+                                     start);
+}
+
+template <ENCODING enc>
+static inline npy_long
+string_rfind(char *str1, int elsize1, char *str2, int elsize2,
+             npy_int64 start, npy_int64 end)
+{
+    Buffer<enc> buf1(str1, elsize1);
+    Buffer<enc> buf2(str2, elsize2);
+
+    npy_int64 len1 = buf1.length();
+    npy_int64 len2 = buf2.length();
+
+    adjust_offsets(&start, &end, len1);
+    if (end - start < len2) {
+        return (npy_long) -1;
+    }
+
+    if (len2 == 1) {
+        npy_ucs4 ch = *buf2;
+        npy_long result = (npy_long) rfindchar(buf1 + start, end - start, ch);
+        if (result == -1) {
+            return (npy_long) -1;
+        } else {
+            return result + (npy_long) start;
+        }
+    }
+
+    return (npy_long) rfindslice(buf1 + start,
+                                 end - start,
+                                 buf2,
+                                 len2,
+                                 start);
 }
 
 
@@ -296,9 +430,138 @@ string_isalpha_loop(PyArrayMethod_Context *context,
 }
 
 
+template<ENCODING enc>
+static int
+string_find_loop(PyArrayMethod_Context *context,
+        char *const data[], npy_intp const dimensions[],
+        npy_intp const strides[], NpyAuxData *NPY_UNUSED(auxdata))
+{
+    int elsize1 = context->descriptors[0]->elsize;
+    int elsize2 = context->descriptors[1]->elsize;
+
+    char *in1 = data[0];
+    char *in2 = data[1];
+    char *in3 = data[2];
+    char *in4 = data[3];
+    char *out = data[4];
+
+    npy_intp N = dimensions[0];
+
+    while (N--) {
+        npy_long idx = string_find<enc>(in1, elsize1, in2, elsize2,
+                                         *(npy_int64 *)in3, *(npy_int64 *)in4);
+        *(npy_long *)out = idx;
+
+        in1 += strides[0];
+        in2 += strides[1];
+        in3 += strides[2];
+        in4 += strides[3];
+        out += strides[4];
+    }
+    return 0;
+}
+
+
+template<ENCODING enc>
+static int
+string_rfind_loop(PyArrayMethod_Context *context,
+        char *const data[], npy_intp const dimensions[],
+        npy_intp const strides[], NpyAuxData *NPY_UNUSED(auxdata))
+{
+    int elsize1 = context->descriptors[0]->elsize;
+    int elsize2 = context->descriptors[1]->elsize;
+
+    char *in1 = data[0];
+    char *in2 = data[1];
+    char *in3 = data[2];
+    char *in4 = data[3];
+    char *out = data[4];
+
+    npy_intp N = dimensions[0];
+
+    while (N--) {
+        npy_long idx = string_rfind<enc>(in1, elsize1, in2, elsize2,
+                                          *(npy_int64 *)in3, *(npy_int64 *)in4);
+        *(npy_long *)out = idx;
+
+        in1 += strides[0];
+        in2 += strides[1];
+        in3 += strides[2];
+        in4 += strides[3];
+        out += strides[4];
+    }
+    return 0;
+}
+
+
+/* Resolve descriptors & promoter functions */
+
+static NPY_CASTING
+string_addition_resolve_descriptors(
+        PyArrayMethodObject *NPY_UNUSED(self),
+        PyArray_DTypeMeta *NPY_UNUSED(dtypes[3]),
+        PyArray_Descr *given_descrs[3],
+        PyArray_Descr *loop_descrs[3],
+        npy_intp *NPY_UNUSED(view_offset))
+{
+    loop_descrs[0] = NPY_DT_CALL_ensure_canonical(given_descrs[0]);
+    if (loop_descrs[0] == NULL) {
+        return _NPY_ERROR_OCCURRED_IN_CAST;
+    }
+
+    loop_descrs[1] = NPY_DT_CALL_ensure_canonical(given_descrs[1]);
+    if (loop_descrs[1] == NULL) {
+        return _NPY_ERROR_OCCURRED_IN_CAST;
+    }
+
+    loop_descrs[2] = PyArray_DescrNew(loop_descrs[0]);
+    if (loop_descrs[2] == NULL) {
+        return _NPY_ERROR_OCCURRED_IN_CAST;
+    }
+    loop_descrs[2]->elsize += loop_descrs[1]->elsize;
+
+    return NPY_NO_CASTING;
+}
+
+static int
+string_find_rfind_promoter(PyUFuncObject *NPY_UNUSED(ufunc),
+        PyArray_DTypeMeta *op_dtypes[], PyArray_DTypeMeta *signature[],
+        PyArray_DTypeMeta *new_op_dtypes[])
+{
+    Py_INCREF(op_dtypes[0]);
+    new_op_dtypes[0] = op_dtypes[0];
+    Py_INCREF(op_dtypes[1]);
+    new_op_dtypes[1] = op_dtypes[1];
+    new_op_dtypes[2] = PyArray_DTypeFromTypeNum(NPY_INT64);
+    new_op_dtypes[3] = PyArray_DTypeFromTypeNum(NPY_INT64);
+    new_op_dtypes[4] = PyArray_DTypeFromTypeNum(NPY_LONG);
+    return 0;
+}
+
+
 /*
  * Machinery to add the string loops to the existing ufuncs.
  */
+
+static int
+install_promoter(PyObject *umath, const char *ufunc_name,
+                 PyObject *promoterinfo)
+{
+    PyObject *name = PyUnicode_FromString(ufunc_name);
+    if (name == nullptr) {
+        return -1;
+    }
+    PyObject *ufunc = PyObject_GetItem(umath, name);
+    Py_DECREF(name);
+    if (ufunc == nullptr) {
+        return -1;
+    }
+
+    int res = PyUFunc_AddLoop((PyUFuncObject *)ufunc, promoterinfo, 0);
+    Py_DECREF(ufunc);
+    return res;
+}
+
 
 /*
  * This function replaces the strided loop with the passed in one,
@@ -319,7 +582,7 @@ add_loop(PyObject *umath, const char *ufunc_name,
     }
     spec->slots[0].pfunc = (void *)loop;
 
-    int res = PyUFunc_AddLoopFromSpec(ufunc, spec);
+    int res = PyUFunc_AddLoopFromSpec_int(ufunc, spec, 1);
     Py_DECREF(ufunc);
     return res;
 }
@@ -401,127 +664,92 @@ init_comparison(PyObject *umath)
 }
 
 
-static NPY_CASTING
-string_addition_resolve_descriptors(
-        PyArrayMethodObject *NPY_UNUSED(self),
-        PyArray_DTypeMeta *NPY_UNUSED(dtypes[3]),
-        PyArray_Descr *given_descrs[3],
-        PyArray_Descr *loop_descrs[3],
-        npy_intp *NPY_UNUSED(view_offset))
-{
-    loop_descrs[0] = NPY_DT_CALL_ensure_canonical(given_descrs[0]);
-    if (loop_descrs[0] == NULL) {
-        return _NPY_ERROR_OCCURRED_IN_CAST;
-    }
-
-    loop_descrs[1] = NPY_DT_CALL_ensure_canonical(given_descrs[1]);
-    if (loop_descrs[1] == NULL) {
-        return _NPY_ERROR_OCCURRED_IN_CAST;
-    }
-
-    loop_descrs[2] = PyArray_DescrNew(loop_descrs[0]);
-    if (loop_descrs[2] == NULL) {
-        return _NPY_ERROR_OCCURRED_IN_CAST;
-    }
-    loop_descrs[2]->elsize += loop_descrs[1]->elsize;
-
-    return NPY_NO_CASTING;
-}
-
-
 static int
-init_add(PyObject *umath)
+init_ufunc(PyObject *umath, const char *name, const char *specname, int nin, int nout,
+           NPY_TYPES *typenums, PyArrayMethod_StridedLoop *stringmethod,
+           PyArrayMethod_StridedLoop unicodemethod,
+           resolve_descriptors_function resolve_descriptors,
+           promoter_function promoter)
 {
     int res = -1;
 
-    PyArray_DTypeMeta *String = PyArray_DTypeFromTypeNum(NPY_STRING);
-    PyArray_DTypeMeta *Unicode = PyArray_DTypeFromTypeNum(NPY_UNICODE);
+    PyArray_DTypeMeta **dtypes = (PyArray_DTypeMeta **) PyMem_Malloc(
+        (nin + nout) * sizeof(PyArray_DTypeMeta *));
+    if (dtypes == NULL) {
+        PyErr_NoMemory();
+        return -1;
+    }
 
-    /* We start with the string loops: */
-    PyArray_DTypeMeta *dtypes[] = {String, String, String};
-    /*
-     * We only have one loop right now, the strided one.  The default type
-     * resolver ensures native byte order/canonical representation.
-     */
-    PyType_Slot slots[] = {
-        {NPY_METH_strided_loop, nullptr},
-        {NPY_METH_resolve_descriptors, (void *) &string_addition_resolve_descriptors},
-        {0, nullptr}
-    };
+    for (int i = 0; i < nin+nout; i++) {
+        dtypes[i] = PyArray_DTypeFromTypeNum(typenums[i]);
+    }
+
+    PyType_Slot slots[3];
+    slots[0] = {NPY_METH_strided_loop, nullptr};
+    slots[2] = {0, nullptr};
+    if (resolve_descriptors != NULL) {
+        slots[1] = {NPY_METH_resolve_descriptors, (void *) resolve_descriptors};
+    } else {
+        slots[1] = {0, nullptr};
+    }
 
     PyArrayMethod_Spec spec = {};
-    spec.name = "templated_add";
-    spec.nin = 2;
-    spec.nout = 1;
+    spec.name = specname;
+    spec.nin = nin;
+    spec.nout = nout;
     spec.dtypes = dtypes;
     spec.slots = slots;
     spec.flags = NPY_METH_NO_FLOATINGPOINT_ERRORS;
 
     /* All String loops */
-    if (add_loop(umath, "add", &spec, string_add_loop<npy_byte>) < 0) {
+    if (add_loop(umath, name, &spec, stringmethod) < 0) {
         goto finish;
     }
 
     /* All Unicode loops */
-    dtypes[0] = Unicode;
-    dtypes[1] = Unicode;
-    dtypes[2] = Unicode;
-    if (add_loop(umath, "add", &spec, string_add_loop<npy_ucs4>) < 0) {
+    for (int i = 0; i < nin+nout; i++) {
+        if (typenums[i] == NPY_STRING) {
+            Py_DECREF(dtypes[i]);
+            dtypes[i] = PyArray_DTypeFromTypeNum(NPY_UNICODE);
+        }
+    }
+    if (add_loop(umath, name, &spec, unicodemethod) < 0) {
         goto finish;
+    }
+
+    if (promoter != NULL) {
+        PyObject *promoter_obj = PyCapsule_New((void *) promoter, "numpy._ufunc_promoter", NULL);
+        if (promoter_obj == NULL) {
+            goto finish;
+        }
+
+        PyObject *dtypes_tuple = PyTuple_New(nin + nout);
+        if (dtypes_tuple == NULL) {
+            Py_DECREF(promoter_obj);
+            goto finish;
+        }
+        for (int i = 0; i < nin + nout; i++) {
+            PyTuple_SET_ITEM(dtypes_tuple, i, Py_None);
+        }
+
+        PyObject *info = PyTuple_Pack(2, dtypes_tuple, promoter_obj);
+        Py_DECREF(dtypes_tuple);
+        Py_DECREF(promoter_obj);
+        if (info == NULL) {
+            goto finish;
+        }
+
+        if (install_promoter(umath, name, info) < 0) {
+            goto finish;
+        }
     }
 
     res = 0;
   finish:
-    Py_DECREF(String);
-    Py_DECREF(Unicode);
-    return res;
-}
-
-
-static int
-init_isalpha(PyObject *umath)
-{
-    int res = -1;
-    /* NOTE: This should receive global symbols? */
-    PyArray_DTypeMeta *String = PyArray_DTypeFromTypeNum(NPY_STRING);
-    PyArray_DTypeMeta *Unicode = PyArray_DTypeFromTypeNum(NPY_UNICODE);
-    PyArray_DTypeMeta *Bool = PyArray_DTypeFromTypeNum(NPY_BOOL);
-
-    /* We start with the string loops: */
-    PyArray_DTypeMeta *dtypes[] = {String, Bool};
-    /*
-     * We only have one loop right now, the strided one.  The default type
-     * resolver ensures native byte order/canonical representation.
-     */
-    PyType_Slot slots[] = {
-        {NPY_METH_strided_loop, nullptr},
-        {0, nullptr}
-    };
-
-    PyArrayMethod_Spec spec = {};
-    spec.name = "templated_string_isalpha";
-    spec.nin = 1;
-    spec.nout = 1;
-    spec.dtypes = dtypes;
-    spec.slots = slots;
-    spec.flags = NPY_METH_NO_FLOATINGPOINT_ERRORS;
-
-    /* All String loops */
-    if (add_loop(umath, "isalpha", &spec, string_isalpha_loop<npy_byte>) < 0) {
-        goto finish;
+    for (int i = 0; i < nin+nout; i++) {
+        Py_DECREF(dtypes[i]);
     }
-
-    /* All Unicode loops */
-    dtypes[0] = Unicode;
-    if (add_loop(umath, "isalpha", &spec, string_isalpha_loop<npy_ucs4>) < 0) {
-        goto finish;
-    }
-
-    res = 0;
-  finish:
-    Py_DECREF(String);
-    Py_DECREF(Unicode);
-    Py_DECREF(Bool);
+    PyMem_Free((void *) dtypes);
     return res;
 }
 
@@ -529,15 +757,41 @@ init_isalpha(PyObject *umath)
 NPY_NO_EXPORT int
 init_string_ufuncs(PyObject *umath)
 {
+    NPY_TYPES dtypes[] = {NPY_STRING, NPY_STRING, NPY_STRING, NPY_STRING, NPY_STRING};
+
     if (init_comparison(umath) < 0) {
         return -1;
     }
 
-    if (init_add(umath) < 0) {
+    if (init_ufunc(umath, "add", "templated_string_add", 2, 1, dtypes,
+                   string_add_loop<npy_byte>, string_add_loop<npy_ucs4>,
+                   string_addition_resolve_descriptors, NULL) < 0) {
         return -1;
     }
 
-    if (init_isalpha(umath) < 0) {
+    dtypes[0] = NPY_STRING;
+    dtypes[1] = NPY_BOOL;
+    if (init_ufunc(umath, "isalpha", "templated_string_isalpha", 1, 1, dtypes,
+                   string_isalpha_loop<npy_byte>, string_isalpha_loop<npy_ucs4>,
+                   NULL, NULL) < 0) {
+        return -1;
+    }
+
+    dtypes[0] = dtypes[1] = NPY_STRING;
+    dtypes[2] = dtypes[3] = NPY_INT64;
+    dtypes[4] = NPY_LONG;
+    if (init_ufunc(umath, "find", "templated_string_find", 4, 1, dtypes,
+                   string_find_loop<ENCODING::ASCII>, string_find_loop<ENCODING::UTF32>,
+                   NULL, string_find_rfind_promoter) < 0) {
+        return -1;
+    }
+
+    dtypes[0] = dtypes[1] = NPY_STRING;
+    dtypes[2] = dtypes[3] = NPY_INT64;
+    dtypes[4] = NPY_LONG;
+    if (init_ufunc(umath, "rfind", "templated_string_rfind", 4, 1, dtypes,
+                   string_rfind_loop<ENCODING::ASCII>, string_rfind_loop<ENCODING::UTF32>,
+                   NULL, string_find_rfind_promoter) < 0) {
         return -1;
     }
 

--- a/numpy/_core/tests/test_ufunc.py
+++ b/numpy/_core/tests/test_ufunc.py
@@ -2901,7 +2901,7 @@ def test_find_non_long_args(dtype):
     end = dtype(len(element))
     arr = np.array([element])
     result = np._core.umath.find(arr, "a", start, end)
-    assert result.dtype == np.dtype("long")
+    assert result.dtype == np.dtype("intp")
     assert result == 0
 
 

--- a/numpy/_core/tests/test_ufunc.py
+++ b/numpy/_core/tests/test_ufunc.py
@@ -2894,6 +2894,28 @@ def test_addition_unicode_inverse_byte_order(order1, order2):
     assert result == 2*element
 
 
+@pytest.mark.parametrize("dtype", [np.int8, np.int16, np.int32, np.int64])
+def test_find_non_long_args(dtype):
+    element = 'abcd'
+    start = dtype(0)
+    end = dtype(len(element))
+    arr = np.array([element])
+    result = np._core.umath.find(arr, "a", start, end)
+    assert result.dtype == np.dtype("long")
+    assert result == 0
+
+
+def test_find_access_past_buffer():
+    # This checks that no read past the string buffer occurs in
+    # string_fastsearch.h. The READC macro is there to check this.
+    # To see it in action, you can redefine READC to just read the
+    # i'th character of the buffer and this test will produce an
+    # 'Invalid read' if run under valgrind.
+    arr = np.array([b'abcd', b'ebcd'])
+    result = np._core.umath.find(arr, b'cde', 0, np.iinfo(np.int64).max)
+    assert np.all(result == -1)
+
+
 class TestLowlevelAPIAccess:
     def test_resolve_dtypes_basic(self):
         # Basic test for dtype resolution:

